File: apps/driver-android/src/test/kotlin/lk/mageride/driver/wallet/WalletTestKit.kt
package lk.mageride.driver.wallet

import lk.mageride.shared.data.models.Page
import lk.mageride.shared.data.models.Timestamp
import lk.mageride.shared.data.models.Ulid
import lk.mageride.shared.data.models.VehicleType
import lk.mageride.shared.data.models.subscription.DailyFeeDayStatus
import lk.mageride.shared.data.models.subscription.TodaysDailyFee
import lk.mageride.shared.data.models.subscription.VoucherDiscountTier
import lk.mageride.shared.data.models.wallet.TransferDirection
import lk.mageride.shared.data.models.wallet.TransferRow
import lk.mageride.shared.data.models.wallet.TransferStatus
import lk.mageride.shared.data.models.wallet.Wallet
import lk.mageride.shared.data.models.wallet.WalletTransaction
import lk.mageride.shared.testing.fixture.Fixtures
import lk.mageride.shared.util.BusinessCalendar
import kotlin.time.ExperimentalTime

/** The driver these tests are signed in as — `FakeApiBackend` routes by operation id, not by path. */
internal const val WALLET_DRIVER_ID: Ulid = Fixtures.DRIVER_ID

/** The other driver in a transfer. A well-formed platform id, because `WalletInput` checks. */
internal const val OTHER_DRIVER_ID: Ulid = Fixtures.RECIPIENT_ID

/** Rs 1,000 in minor units — the denomination US-9.19 gives its one worked rate for. */
internal const val ONE_THOUSAND: Long = 100_000L

/**
 * A wallet holding [balanceMinor], with [debtMinor] of it already owed.
 *
 * The two are separate because every affordability question in this cluster asks about the
 * **spendable** figure (D-05): a driver holding Rs 300 who owes Rs 200 can spend Rs 100.
 */
internal fun wallet(balanceMinor: Long, debtMinor: Long = 0L): Wallet = Wallet(
    userId = WALLET_DRIVER_ID,
    balanceMinor = balanceMinor,
    availableMinor = balanceMinor - debtMinor,
    outstandingDebtMinor = debtMinor,
)

/** Today's fee row for a three-wheeler at D5' §2.1's Rs 100 tier. */
@OptIn(ExperimentalTime::class)
internal fun todaysFee(
    paid: Boolean = false,
    tripsToday: Int = 0,
    rateMinor: Long = 10_000L,
    vehicleType: VehicleType = VehicleType.THREE_WHEELER,
): TodaysDailyFee = TodaysDailyFee(
    vehicleType = vehicleType,
    vehicleId = Fixtures.VEHICLE_ID,
    dailyRateMinor = rateMinor,
    status = if (paid) DailyFeeDayStatus.PAID else DailyFeeDayStatus.UNPAID,
    deductedMinor = if (paid) rateMinor else 0L,
    tripsToday = tripsToday,
    // The free first trip is spent by the charge, so "paid" and "still free" never coexist.
    firstTripFree = !paid,
    feeDate = BusinessCalendar.businessDate(Fixtures.NOW),
)

/** One voucher tier. `1000` bps is the 10% US-9.19 works through. */
internal fun tier(denominationMinor: Long, discountBps: Int, active: Boolean = true): VoucherDiscountTier =
    VoucherDiscountTier(denominationMinor = denominationMinor, discountBps = discountBps, active = active)

/**
 * One credit-transfer row, from the reading driver's point of view.
 *
 * The counterparty and the timestamp are fixed: no rule in this cluster reads either, and what a
 * transfer test is about is the amount, the direction and the status.
 */
@OptIn(ExperimentalTime::class)
internal fun transfer(
    transferId: Ulid,
    amountMinor: Long,
    direction: TransferDirection = TransferDirection.RECEIVED,
    status: TransferStatus = TransferStatus.PENDING,
): TransferRow = TransferRow(
    transferId = transferId,
    counterpartyDriverId = OTHER_DRIVER_ID,
    counterpartyName = "S. Bandara",
    amountMinor = amountMinor,
    direction = direction,
    status = status,
    createdAt = Fixtures.NOW,
)

/** One ledger line. [amountMinor] is **signed** — a debit is negative (D3' §0). */
@OptIn(ExperimentalTime::class)
internal fun ledgerLine(
    entryId: Ulid,
    kind: String,
    amountMinor: Long,
    balanceAfterMinor: Long = 0L,
    at: Timestamp = Fixtures.NOW,
): WalletTransaction = WalletTransaction(
    transactionId = entryId,
    entryId = entryId,
    kind = kind,
    amountMinor = amountMinor,
    balanceAfterMinor = balanceAfterMinor,
    occurredAt = at,
)

/** A one-page answer for any of the paged wallet reads. */
internal fun <T> onePage(vararg rows: T): Page<T> = Page(items = rows.toList())

/** [WalletPreferences] in memory — the production one is `SharedPreferences` (US-9.9's local line). */
internal class FakeWalletPreferences(override var lowBalanceThresholdMinor: Long? = null) : WalletPreferences

/**
 * [PaymentHandoff] that records what it was asked to open.
 *
 * [handled] is what makes AL-15's fallback testable: the rule is *"try the deep link, show the code
 * when nothing opened it"*, and the only way to reach the second half is a handset that refuses.
 */
internal class FakePaymentHandoff(var handled: Boolean = true) : PaymentHandoff {

    /** Every URL offered, oldest first. */
    val opened: MutableList<String> = mutableListOf()

    override fun open(url: String): Boolean {
        opened += url
        return handled
    }
}

/** [StatementExporter] that keeps the bytes instead of writing them. */
internal class FakeStatementExporter(var handled: Boolean = true) : StatementExporter {

    /** What the last successful export was asked to share. */
    var lastFileName: String? = null
    var lastFormat: StatementFormat? = null
    var lastBytes: ByteArray? = null

    override fun share(fileName: String, format: StatementFormat, bytes: ByteArray): Boolean {
        lastFileName = fileName
        lastFormat = format
        lastBytes = bytes
        return handled
    }
}

File: apps/driver-ios/DriverAppTests/WalletTestKit.swift
import Foundation
import MageRideShared

@testable import DriverApp

/// The fakes and fixtures cluster 3's money screens are driven by.
///
/// Same rule as ``DashboardTestKit`` and ``JobsTestKit``: every seam here is a **Swift protocol**,
/// because the Kotlin types behind them are interfaces Swift cannot stand in for. The DTOs underneath
/// are the real shared ones — a fixture is built with the same initialiser the gateway's response
/// deserialises into, so a contract change fails these tests rather than a driver's phone.
///
/// **Every repository is recorded, not counted.** Half of what C091's rules say is *which* call ran: a
/// voucher goes to `purchaseVoucher` and never to `topUp`, an approval re-reads the balance, and a
/// rejected request never joins the history. A counter cannot say any of that.

/// A second driver, so a self-transfer and a transfer to somebody else are different tests.
// Crockford base32 has no `O` and no `L` — it spells them `0` and `1` — so "HOLDER" is `H01DER`
// here. The old spelling failed `PlatformId.isValid`, which is the rule this id exists to satisfy.
let testHolderId = "01JH01DER00000000000000001"

let testTransferId = "01JTRANSFER0000000000000A"

/// `GET /v1/wallet/{userId}` — a wallet with money in it and no debt.
func driverWallet(
    balanceMinor: Int64 = 124_000,
    availableMinor: Int64? = nil,
    outstandingDebtMinor: Int64? = nil
) -> Wallet {
    Wallet(
        userId: testDriverId,
        balanceMinor: balanceMinor,
        availableMinor: availableMinor ?? balanceMinor,
        outstandingDebtMinor: outstandingDebtMinor.map { KotlinLong(value: $0) },
        currency: Currency.lkr,
        updatedAt: nil
    )
}

/// `GET /v1/fees/{driverId}/today`.
func todaysDailyFee(
    vehicleType: VehicleType = VehicleType.threeWheeler,
    dailyRateMinor: Int64 = 10_000,
    status: DailyFeeDayStatus = DailyFeeDayStatus.unpaid,
    tripsToday: Int32 = 0,
    firstTripFree: Bool = true
) -> TodaysDailyFee {
    TodaysDailyFee(
        vehicleType: vehicleType,
        vehicleId: "01JVEHICLE000000000000001",
        dailyRateMinor: dailyRateMinor,
        status: status,
        deductedMinor: nil,
        tripsToday: tripsToday,
        firstTripFree: firstTripFree,
        feeDate: IosBusinessDateKt.colomboBusinessDateNow(),
        feeDateTzAt: nil
    )
}

/// `GET /v1/fees/rates` — D5' §2.1's own tiers, as a schedule.
func feeSchedule(threeWheelerMinor: Int64 = 10_000) -> DailyFeeSchedule {
    DailyFeeSchedule(rates: [
        DailyFeeRate(
            vehicleType: VehicleType.threeWheeler,
            dailyFeeMinor: threeWheelerMinor,
            mode: ServiceMode.c,
            currency: Currency.lkr
        ),
    ])
}

/// One rung of the bulk-voucher ladder (`billing.voucher_discount_tiers`).
func voucherTier(denominationMinor: Int64, discountBps: Int32, active: Bool = true) -> VoucherDiscountTier {
    VoucherDiscountTier(
        denominationMinor: denominationMinor,
        discountBps: discountBps,
        active: active,
        updatedAt: nil
    )
}

/// The wireframe's own ladder — `1k +10%`, `2k +12%`, `5k +15%`.
let testVoucherTiers: [VoucherDiscountTier] = [
    voucherTier(denominationMinor: 100_000, discountBps: 1_000),
    voucherTier(denominationMinor: 200_000, discountBps: 1_200),
    voucherTier(denominationMinor: 500_000, discountBps: 1_500),
]

func topup(
    topupId: String = "01JTOPUP0000000000000001",
    state: TopupState = TopupState.pending,
    amountMinor: Int64 = 200_000,
    redirectUrl: String? = "https://checkout.onepay.lk/session/abc",
    paymentLink: String? = nil,
    qrPayload: String? = nil
) -> Topup {
    Topup(
        topupId: topupId,
        state: state,
        amountMinor: amountMinor,
        currency: Currency.lkr,
        redirectUrl: redirectUrl,
        sessionToken: nil,
        paymentLink: paymentLink,
        qrPayload: qrPayload
    )
}

/// `POST /v1/vouchers/purchase` — `creditedMinor` is always the face value (`ck_voucher_credit_full`).
func voucherPurchase(
    denominationMinor: Int64 = 100_000,
    discountBps: Int32 = 1_000,
    paidMinor: Int64 = 90_000,
    redirectUrl: String? = "https://checkout.onepay.lk/session/voucher",
    qrPayload: String? = nil
) -> VoucherPurchase {
    VoucherPurchase(
        purchaseId: "01JVOUCHER00000000000001",
        denominationMinor: denominationMinor,
        discountBpsApplied: discountBps,
        paidMinor: paidMinor,
        creditedMinor: denominationMinor,
        currency: Currency.lkr,
        redirectUrl: redirectUrl,
        qrPayload: qrPayload
    )
}

func transferRow(
    transferId: String = testTransferId,
    counterpartyDriverId: String = testHolderId,
    counterpartyName: String? = "S. Bandara",
    amountMinor: Int64 = 100_000,
    direction: TransferDirection = TransferDirection.received,
    status: TransferStatus = TransferStatus.pending
) -> TransferRow {
    TransferRow(
        transferId: transferId,
        counterpartyDriverId: counterpartyDriverId,
        counterpartyName: counterpartyName,
        amountMinor: amountMinor,
        currency: Currency.lkr,
        direction: direction,
        status: status,
        createdAt: timestamp(testNow)
    )
}

/// One ledger line. `amountMinor` is **signed** — a debit is negative (D3' §0's own exemption).
func walletTransaction(
    entryId: String = "01JENTRY00000000000000001",
    kind: String = LedgerKinds.topup,
    amountMinor: Int64 = 200_000,
    balanceAfterMinor: Int64 = 324_000,
    reference: String? = nil,
    occurredAt: Date = testNow
) -> WalletTransaction {
    WalletTransaction(
        transactionId: "tx-" + entryId,
        entryId: entryId,
        kind: kind,
        amountMinor: amountMinor,
        currency: Currency.lkr,
        balanceAfterMinor: balanceAfterMinor,
        reference: reference,
        occurredAt: timestamp(occurredAt)
    )
}

// MARK: - Seams

/// ``WalletRepository`` with no gateway.
final class FakeWalletRepository: WalletRepository {

    var standing = WalletFeeStanding(wallet: driverWallet(), dailyFee: todaysDailyFee(), schedule: feeSchedule())
    var balances: [WalletStanding] = []
    var lines: [WalletTransaction] = []
    var statementBytes = Data("date,amount\n".utf8)
    var nextBalanceFailure: Error?
    var nextTransactionsFailure: Error?
    var nextStatementFailure: Error?

    private(set) var standingReads = 0
    private(set) var balanceReads = 0
    private(set) var transactionReads: [(from: BusinessDate?, to: BusinessDate?)] = []
    private(set) var statementReads: [(format: StatementFormat, from: BusinessDate?, to: BusinessDate?)] = []

    func standing(driverId: String) async -> WalletFeeStanding {
        standingReads += 1
        return standing
    }

    /// Answers `balances` in order and then repeats the last, so a test can say *"the balance was
    /// Rs 300 and after the approval it is Rs 200"* without programming a closure.
    func balance(driverId: String) async throws -> WalletStanding {
        balanceReads += 1
        if let failure = nextBalanceFailure {
            nextBalanceFailure = nil
            throw failure
        }
        guard !balances.isEmpty else { return WalletStanding.companion.of(wallet: driverWallet()) }
        return balances[min(balanceReads - 1, balances.count - 1)]
    }

    func transactions(driverId: String, from: BusinessDate?, to: BusinessDate?) async throws -> [WalletTransaction] {
        transactionReads.append((from, to))
        if let failure = nextTransactionsFailure {
            nextTransactionsFailure = nil
            throw failure
        }
        return lines
    }

    func statement(
        driverId: String,
        format: StatementFormat,
        from: BusinessDate?,
        to: BusinessDate?
    ) async throws -> Data {
        statementReads.append((format, from, to))
        if let failure = nextStatementFailure {
            nextStatementFailure = nil
            throw failure
        }
        return statementBytes
    }
}

/// ``CreditTransferRepository`` with no gateway.
final class FakeCreditTransferRepository: CreditTransferRepository {

    var pendingRows: [TransferRow] = []
    var historyRows: [TransferRow] = []
    var nextRequestFailure: Error?
    var nextSendFailure: Error?
    var nextPendingFailure: Error?
    var nextDecisionFailure: Error?

    /// What the decision answers with. Defaults to the row it was given, approved.
    var decisionResult: TransferRow?

    private(set) var requests: [(holderDriverId: String, amountMinor: Int64)] = []
    private(set) var sends: [(recipientDriverId: String, amountMinor: Int64)] = []
    private(set) var historyReads: [TransferDirectionFilter?] = []
    private(set) var approvals: [String] = []
    private(set) var rejections: [String] = []

    func request(holderDriverId: String, amountMinor: Int64) async throws -> TransferRow {
        requests.append((holderDriverId, amountMinor))
        try throwIf(&nextRequestFailure)
        return transferRow(counterpartyDriverId: holderDriverId, amountMinor: amountMinor)
    }

    func send(recipientDriverId: String, amountMinor: Int64) async throws -> TransferRow {
        sends.append((recipientDriverId, amountMinor))
        try throwIf(&nextSendFailure)
        return transferRow(
            counterpartyDriverId: recipientDriverId,
            amountMinor: amountMinor,
            direction: TransferDirection.sent,
            status: TransferStatus.approved
        )
    }

    func pending() async throws -> [TransferRow] {
        try throwIf(&nextPendingFailure)
        return pendingRows
    }

    func history(driverId: String, direction: TransferDirectionFilter?) async throws -> [TransferRow] {
        historyReads.append(direction)
        return historyRows
    }

    func approve(transferId: String) async throws -> TransferRow {
        approvals.append(transferId)
        try throwIf(&nextDecisionFailure)
        return decisionResult ?? transferRow(transferId: transferId, status: TransferStatus.approved)
    }

    func reject(transferId: String) async throws -> TransferRow {
        rejections.append(transferId)
        try throwIf(&nextDecisionFailure)
        return decisionResult ?? transferRow(transferId: transferId, status: TransferStatus.rejected)
    }

    private func throwIf(_ failure: inout Error?) throws {
        guard let programmed = failure else { return }
        failure = nil
        throw programmed
    }
}

/// ``TopUpRepository`` with no gateway.
final class FakeTopUpRepository: TopUpRepository {

    var tiers: [VoucherDiscountTier] = testVoucherTiers
    var startedTopup = topup()
    /// The sessions `topUpState` answers, in order; the last one repeats.
    var polledStates: [Topup] = []
    var purchase = voucherPurchase()
    var nextTiersFailure: Error?
    var nextTopUpFailure: Error?
    var nextPurchaseFailure: Error?

    private(set) var tierReads = 0
    private(set) var topUps: [(method: TopupMethod, amountMinor: Int64)] = []
    private(set) var purchases: [(denominationMinor: Int64, method: TopupMethod)] = []
    private(set) var pollReads = 0

    func voucherTiers() async throws -> VoucherCatalogue {
        tierReads += 1
        try throwIf(&nextTiersFailure)
        return VoucherCatalogue(tiers: tiers)
    }

    func topUp(method: TopupMethod, amountMinor: Int64) async throws -> Topup {
        topUps.append((method, amountMinor))
        try throwIf(&nextTopUpFailure)
        return startedTopup
    }

    func topUpState(topupId: String) async throws -> Topup {
        pollReads += 1
        guard !polledStates.isEmpty else { return topup(state: TopupState.succeeded) }
        return polledStates[min(pollReads - 1, polledStates.count - 1)]
    }

    func buyVoucher(denominationMinor: Int64, method: TopupMethod) async throws -> VoucherPurchase {
        purchases.append((denominationMinor, method))
        try throwIf(&nextPurchaseFailure)
        return purchase
    }

    private func throwIf(_ failure: inout Error?) throws {
        guard let programmed = failure else { return }
        failure = nil
        throw programmed
    }
}

/// ``PaymentHandoff`` with no `UIApplication`.
@MainActor
final class FakePaymentHandoff: PaymentHandoff {

    /// Whether a bank app claims the link. `false` is AL-15's fallback condition.
    var opens = true

    private(set) var openedUrls: [String] = []

    func openBankApp(_ url: String) async -> Bool {
        openedUrls.append(url)
        return opens
    }
}

/// ``WalletPreferences`` in memory.
final class FakeWalletPreferences: WalletPreferences {

    var lowBalanceThresholdMinor: Int64?
}

/// ``StatementExporter`` with no filesystem.
final class FakeStatementExporter: StatementExporter {

    /// `nil` is the "could not be written" branch.
    var url: URL? = URL(fileURLWithPath: "/tmp/mageride-wallet-all-all.csv")

    private(set) var writes: [(fileName: String, bytes: Data)] = []

    func write(fileName: String, bytes: Data) -> URL? {
        writes.append((fileName, bytes))
        return url
    }
}

/// A failure that is not a `MageRideError`, so it resolves to the shell's generic copy.
struct TestWalletFailure: Error {}
